## Analysing Car Listings on eBay  Kleinanzeigen.
The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The original dataset isn't available anymore. The data was provided by dataquest and it contains 50,000 rows.

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `odometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

# Importing the Pandas Library

The pandas library will be used for analysing this data. The data is going to be parsed using the pd.read_csv method and then the first 5 rows will be printed using the pd.head() method as well as the general info using the pd.info() method.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

# Cleaning The columns:
The columns are in camelcase instead of python's preffered snakecase. the columns are going to  be chnaged to snake_case which will make it easier for indexing. Some column names will be completely changed. These columns are:

`yearOfRegistration` to `registration_year`

`monthOfRegistration` to `registration_month`

`notRepairedDamage` to `unrepaired_damage`

`dateCreated` to `ad_created`

In [5]:
columns = autos.columns
print(columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [6]:
autos.columns =['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_of_pictures', 'postal_code',
       'last_seen']

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Initial Exploration/ Data Cleaning:
The pd.describe() method will be used here to give an overview of data in each column. The include all argument is used to also get information on non-numeric data. With, it'll be easy to figure out how to clean the data

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

With further inspection of our data frame, we can see that the columns such as the offer_type, seller and no_of_pictures all have the same values and so they are not necessary for our analysis. We are going to drop those columns using the pd.drop() method.

In [9]:
autos.drop(['offer_type', 'seller', 'no_of_pictures'], axis=1, inplace=True)

From the inspection of the dataframe, we can see that our price and odometer columns are type object which means that they are strings instead of integers or floats. We are going to remove the non numeric characters from these coumns and then convert them to integers. We are also going to rename our odometer column to odometer_km since we are getting rid of the km symbol.

In [10]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)
autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '').astype(int)
autos.rename({'odometer': 'odometer_km'}, axis =1,  inplace = True)

# Exploring Price and Odometer Column

In [11]:
autos['price'].unique().shape

(2357,)

In [12]:
print(autos['price'].describe())

autos['price'].value_counts().head()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

In [13]:
autos['price'].sort_values(ascending=False).head(10)

39705    99999999
42221    27322222
39377    12345678
47598    12345678
27371    12345678
2897     11111111
24384    11111111
11137    10000000
47634     3890000
7814      1300000
Name: price, dtype: int64

In [14]:
autos['price'].sort_values().head(10)

18089    0
43923    0
15225    0
29499    0
8445     0
38832    0
43925    0
8438     0
35821    0
31332    0
Name: price, dtype: int64

In [15]:
print(autos['price'].value_counts().sort_index(ascending=False).head(10))

print(autos['price'].value_counts().sort_index().head(10))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64


After exploring our price column we can find out that there are some outliers. For example, the most expensive car listed is $100,000,00 while the least expensive cars have no prices on them. Since the data we are analysing is for used cars, we are going to set the upper limit for the price at $300,000 and for the lower limit, we are going to use $1 as auctions from Ebay can start as low as that price.

In [16]:
autos = autos[autos['price'].between(1, 300000)]

In [17]:
autos['odometer_km'].unique().shape

(13,)

In [18]:
print(autos['odometer_km'].describe())

print(autos['odometer_km'].value_counts())

count     48563.000000
mean     125774.766798
std       39782.786221
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        835
40000       815
30000       780
20000       761
10000       253
Name: odometer_km, dtype: int64


In [19]:
autos['odometer_km']


0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Length: 48563, dtype: int64

# ExploringThe Columns With Dates.
This includes the 'date_crawled', 'ad_created', 'last_seen', 'registration_month' and 'registration_year' columns.
The data for the registration month and year are stored as integers while the data for the resst are stored as strings.


In [20]:
autos.loc[0:5, ['date_crawled', 'ad_created', 'last_seen']]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21

In [21]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025328
2016-03-06    0.014044
2016-03-07    0.036015
2016-03-08    0.033297
2016-03-09    0.033091
2016-03-10    0.032185
2016-03-11    0.032576
2016-03-12    0.036921
2016-03-13    0.015670
2016-03-14    0.036550
2016-03-15    0.034285
2016-03-16    0.029611
2016-03-17    0.031629
2016-03-18    0.012911
2016-03-19    0.034780
2016-03-20    0.037889
2016-03-21    0.037374
2016-03-22    0.032988
2016-03-23    0.032226
2016-03-24    0.029343
2016-03-25    0.031608
2016-03-26    0.032206
2016-03-27    0.031073
2016-03-28    0.034862
2016-03-29    0.034100
2016-03-30    0.033668
2016-03-31    0.031835
2016-04-01    0.033688
2016-04-02    0.035480
2016-04-03    0.038610
2016-04-04    0.036489
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The above data shows that the site was crawled mainly between the months of March 2016 and April 2016.

In [22]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038857
2016-04-04    0.036859
2016-04-05    0.011820
2016-04-06    0.003254
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

The data above shows the period when an ad eas created. From this data, it shows that some of the ads were created as far back as June of 2015 and the most recent ad was created April of 2016.

In [23]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009596
2016-03-10    0.010667
2016-03-11    0.012376
2016-03-12    0.023784
2016-03-13    0.008896
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016453
2016-03-17    0.028087
2016-03-18    0.007351
2016-03-19    0.015835
2016-03-20    0.020654
2016-03-21    0.020633
2016-03-22    0.021374
2016-03-23    0.018533
2016-03-24    0.019768
2016-03-25    0.019212
2016-03-26    0.016803
2016-03-27    0.015629
2016-03-28    0.020860
2016-03-29    0.022342
2016-03-30    0.024772
2016-03-31    0.023784
2016-04-01    0.022795
2016-04-02    0.024916
2016-04-03    0.025204
2016-04-04    0.024484
2016-04-05    0.124766
2016-04-06    0.221815
2016-04-07    0.131932
Name: last_seen, dtype: float64

The last_seen data tells when last a listing was seen by the crawler. This can be as a result of the cars being sold.

In [24]:
autos['registration_year'].describe()

count    48563.000000
mean      2004.755040
std         88.645692
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Using the describe() method to look at the registration year, it shows that the minimum registration year is the year 1000 and that's centuries before the first cars were invented. The maximum registration year is shown to be 9999 which is too far in the future.

# Fixing Incorrect Registration Year

In [25]:
(autos['registration_year'].between(1930, 2016)).sum()/autos.shape[0]

0.9610608899779668

I've decided to use 1930 as my lower bound because that was the decade the beetle went into production. From our data, it shows that about 96% of our data falls between the registration year of 1930-2016. Only about 4% falls outside this range. Data outside of this range are going to be removed.

In [26]:
autos = autos[autos['registration_year'].between(1930, 2016)]

In [27]:
autos['registration_year'].value_counts(normalize=True)

2000    0.067621
2005    0.062907
1999    0.062071
2004    0.057915
2003    0.057829
          ...   
1939    0.000021
1943    0.000021
1948    0.000021
1953    0.000021
1952    0.000021
Name: registration_year, Length: 75, dtype: float64

The data above showed that the 5 most common years from the registration year are 2000,2005, 1999,2004 and 2003 with each of these years making up at least 6% of our data. While the least common years are 1939, 1943,1948, 1953, 1952 and each of them make up less than 0.1% of our data

# Exploring the mean price and milage by brand

In [28]:
brands = autos['brand'].unique()

In [29]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211304
bmw               0.110045
opel              0.107559
mercedes_benz     0.096482
audi              0.086583
ford              0.069892
renault           0.047137
peugeot           0.029847
fiat              0.025647
seat              0.018276
skoda             0.016412
nissan            0.015277
mazda             0.015191
smart             0.014163
citroen           0.014013
toyota            0.012706
hyundai           0.010027
sonstige_autos    0.009770
volvo             0.009149
mini              0.008763
mitsubishi        0.008228
honda             0.007842
kia               0.007071
alfa_romeo        0.006642
porsche           0.006106
suzuki            0.005935
chevrolet         0.005699
chrysler          0.003514
dacia             0.002635
daihatsu          0.002507
jeep              0.002271
subaru            0.002143
land_rover        0.002100
saab              0.001650
jaguar            0.001564
daewoo            0.001500
trabant           0.001371
r

In [30]:
brand_counts = autos['brand'].value_counts(normalize=True)

In [31]:
top_10_brands =  brand_counts[brand_counts > .016].index
print(top_10_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda'],
      dtype='object')


I'ved decided to look at the data of the top 10 most common brands in our data. German cars such as Volkswagen, BMW, Opel, Mercedes Benz and Audi make up the top 5 cars in our data. With foreign brands such as Ford, Renault and Peugeot completing the top 10

In [32]:
top_10_mean_price = {}
for b in top_10_brands:
    brand_needed = autos[autos['brand'] == b]
    mean_price = brand_needed['price'].mean()
    top_10_mean_price[b] = int(mean_price)
    
print(top_10_mean_price)

{'volkswagen': 5402, 'bmw': 8332, 'opel': 2976, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3745, 'renault': 2475, 'peugeot': 3094, 'fiat': 2813, 'seat': 4397, 'skoda': 6368}


After fetching the mean prices for the top 10 brands. It shows that Audi the cars are the most expensive on average, with the Renault cars being the least expensive on average. On average, the German cars are more expensive than their foreign counterpart.

In [34]:
top_10_mean_milage ={}
for b in top_10_brands:
    brand_needed = autos[autos['brand'] == b]
    mean_milage = brand_needed['odometer_km'].mean()
    top_10_mean_milage[b] = int(mean_milage)
print(top_10_mean_milage)   

{'volkswagen': 128707, 'bmw': 132597, 'opel': 129311, 'mercedes_benz': 130788, 'audi': 129157, 'ford': 124302, 'renault': 128127, 'peugeot': 127153, 'fiat': 117121, 'seat': 121131, 'skoda': 110848}


In [35]:
top10mp_series = pd.Series(top_10_mean_price)
print(top10mp_series)

volkswagen       5402
bmw              8332
opel             2976
mercedes_benz    8628
audi             9336
ford             3745
renault          2475
peugeot          3094
fiat             2813
seat             4397
skoda            6368
dtype: int64


In [36]:
top10mm_series = pd.Series(top_10_mean_milage)
print(top10mm_series)

volkswagen       128707
bmw              132597
opel             129311
mercedes_benz    130788
audi             129157
ford             124302
renault          128127
peugeot          127153
fiat             117121
seat             121131
skoda            110848
dtype: int64


In [38]:
top_10_cars = pd.DataFrame(top10mp_series, columns=['mean_price'])
top_10_cars

mean_price
volkswagen           5402
bmw                  8332
opel                 2976
mercedes_benz        8628
audi                 9336
ford                 3745
renault              2475
peugeot              3094
fiat                 2813
seat                 4397
skoda                6368

In [42]:
top_10_cars['mean_milage'] = top10mm_series
top_10_cars.sort_values('mean_milage', ascending=False, inplace=True)
top_10_cars

mean_price  mean_milage
bmw                  8332       132597
mercedes_benz        8628       130788
opel                 2976       129311
audi                 9336       129157
volkswagen           5402       128707
renault              2475       128127
peugeot              3094       127153
ford                 3745       124302
seat                 4397       121131
fiat                 2813       117121
skoda                6368       110848

# Conclusion:

From the dataframe above, it shows all of the cars have mean milage of over 100,000km. Of the cars with the top 4 highest miage, 3 of those cars are also the 